Yosemite/Yosemite/Stores/OrderStore.swift

```diff
@@ -459,15 +463,22 @@ private extension OrderStore {
     /// Updates an order to be considered as paid locally, for use cases where the payment is captured in the
     /// app to prevent from multiple charging for the same order after subsequent failures (e.g. Interac in Canada).
     ///
-    func markOrderAsPaidLocally(siteID: Int64, orderID: Int64, datePaid: Date, onCompletion: (Result<Order, Error>) -> Void) {
-        let storage = storageManager.viewStorage
-        guard let order = storage.loadOrder(siteID: siteID, orderID: orderID) else {
-            return onCompletion(.failure(MarkOrderAsPaidLocallyError.orderNotFoundInStorage))
-        }
-        order.datePaid = datePaid
-        order.statusKey = OrderStatusEnum.processing.rawValue
-        storage.saveIfNeeded()
-        onCompletion(.success(order.toReadOnly()))
+    func markOrderAsPaidLocally(siteID: Int64, orderID: Int64, datePaid: Date, onCompletion: @escaping (Result<Order, Error>) -> Void) {
+        storageManager.performAndSave({ storage in
+            guard let order = storage.loadOrder(siteID: siteID, orderID: orderID) else {
+                throw MarkOrderAsPaidLocallyError.orderNotFoundInStorage
+            }
+            order.datePaid = datePaid
+            order.statusKey = OrderStatusEnum.processing.rawValue
+            return order
+        }, completion: { result in
+            switch result {
+            case .success(let storageOrder):
+                onCompletion(.success(storageOrder.toReadOnly()))
```

Comment:
The crash happens here, up until getting the completion result we're good, but when we convert to Yosemite.Order then we find the crash. I suspect because threading issues, since we cannot see any log regarding the assertion of the type conversion in the console.

I've tried several things:

1. Does this really need to happen in the main queue? Maybe not, but it also crashes if we move it to a background queue, so nothing we can do through this path.

2. Can we force the type conversion on the same thread? I copied `toReadOnly()` to be executed in the main thread , then wrapped the result type as:
```
order.toReadOnlyOnMain(onCompletion: { storageSyncedOrder in
	onCompletion(.success(storageSyncedOrder))
})
``` 
But this didn't help, it continue crashing at this point.

3. If type conversion fails and we do not know why, do we even need the conversion? I don't think so: We're not using the returning Order type anywhere where `markOrderAsPaidLocally` is called as far as I can see, so I changed the implementation to return `onCompletion: @escaping (Result<Void, Error>) -> Void)` instead:

This seems to work, there is no crash, and everything seems to work as expected. That said, I'll loop the rest of payment folks to see if I missed anything here. If we can change the implementation of the method then this PR should be good to go as well 👍

Reply:
PR is here: https://github.com/woocommerce/woocommerce-ios/pull/14113, from my testing on your branch, once that one is merged then we shouldn't see the crash anymore when collecting an Interac card present payment, but needs to be tested to be sure.

Reply:
Thank you @iamgabrielma for testing! The crash could be due to the concurrency debug check when passing the storage order from the background to the main thread.

I added a commit c8dd2b16622d80a8414cc7aa6e1c34f5de4c35db to return the readonly type, and an assert in 512da700fa0370a56171ac58b39999876a775f87 to check in runtime to ensure the generic type T is not `NSManagedObject`. The reason for the crash could be that NSManagedObject instances are not thread-safe, so we should not pass them between different contexts.

Let me know what you think about this!